src/Workspaces/Core/Portable/FlowAnalysis/CustomDataFlowAnalysis.cs

```diff
@@ -0,0 +1,394 @@
+                                DispatchExceptionThroughCatches(enclosing, startAt: 1);
+                                break;
+
+                            default:
+                                throw ExceptionUtilities.UnexpectedValue(enclosing.Kind);
+                        }
+                    }
+                    else if (fromRegion.Kind == ControlFlowRegionKind.Filter)
+                    {
+                        // If filter throws, dispatch is resumed at the next catch with an original exception
+                        Debug.Assert(enclosing.Kind == ControlFlowRegionKind.FilterAndHandler);
+                        ControlFlowRegion tryAndCatch = enclosing.EnclosingRegion;
+                        Debug.Assert(tryAndCatch.Kind == ControlFlowRegionKind.TryAndCatch);
+
+                        int index = tryAndCatch.NestedRegions.IndexOf(enclosing, startIndex: 1);
+
+                        if (index > 0)
+                        {
+                            DispatchExceptionThroughCatches(tryAndCatch, startAt: index + 1);
+                            fromRegion = tryAndCatch;
+                            continue;
+                        }
+
+                        throw ExceptionUtilities.Unreachable;
+                    }
+
+                    fromRegion = enclosing;
+                }
+                while (fromRegion != null);
+            }
+
+            void DispatchExceptionThroughCatches(ControlFlowRegion tryAndCatch, int startAt)
+            {
+                // For simplicity, we do not try to figure out whether a catch clause definitely
+                // handles all exceptions.
+
+                Debug.Assert(tryAndCatch.Kind == ControlFlowRegionKind.TryAndCatch);
+                Debug.Assert(startAt > 0);
+                Debug.Assert(startAt <= tryAndCatch.NestedRegions.Length);
+
+                for (int i = startAt; i < tryAndCatch.NestedRegions.Length; i++)
+                {
+                    ControlFlowRegion @catch = tryAndCatch.NestedRegions[i];
+
+                    switch (@catch.Kind)
+                    {
+                        case ControlFlowRegionKind.Catch:
+                            toVisit.Add(@catch.FirstBlockOrdinal);
+                            break;
+
+                        case ControlFlowRegionKind.FilterAndHandler:
+                            BasicBlock entryBlock = blocks[@catch.FirstBlockOrdinal];
+                            Debug.Assert(@catch.NestedRegions[0].Kind == ControlFlowRegionKind.Filter);
+                            Debug.Assert(entryBlock.Ordinal == @catch.NestedRegions[0].FirstBlockOrdinal);
+
+                            toVisit.Add(entryBlock.Ordinal);
+                            break;
+
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(@catch.Kind);
+                    }
+                }
+            }
+        }
+    }
```

Comment:
skipping this file.  again, very surprised by the exceptionally high level of complexity, combined with almost no comments practically impossible to review this with confidence.

Reply:
This is a port of the compiler's control flow walker for reachability analysis (which was initially shared, but now has been forked). There are very few changes here and I wouldn't like to diverge with compiler's implementation/style/terminology to keep this file as much in sync with compiler's code as possible.

---
In reply to: [230606659](https://github.com/dotnet/roslyn/pull/30777#discussion_r230606659) [](ancestors = 230606659)